cli/democluster: TestTransientClusterSimulateLatencies failed

cli/democluster.TestTransientClusterSimulateLatencies [failed](https://tanzanite.cluster.engflow.com/invocations/default/3dcf21d5-8aff-4b7e-bdf3-a3bf485d3341?testReportRun=23&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cvY2xpL2RlbW9jbHVzdGVyOmRlbW9jbHVzdGVyX3Rlc3Q=) on release-24.1.0-rc @ [9d77dad6162bc528e725d0c57ec01c6a07d2aea2](https://github.com/cockroachdb/cockroach/commits/9d77dad6162bc528e725d0c57ec01c6a07d2aea2):

Fatal error:

```
panic: test timed out after 4m57s
running tests:
	TestTransientClusterSimulateLatencies (4m57s)

```
Stack: 

```
goroutine 166740 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
* crypto/tls.(*Conn).readRecordOrCCS(0xc012b73508, 0x0)
* 	GOROOT/src/crypto/tls/conn.go:626 +0x3cf
* crypto/tls.(*Conn).readRecord(...)
* 	GOROOT/src/crypto/tls/conn.go:588
* crypto/tls.(*Conn).Read(0xc012b73508, {0xc0139e0000, 0x8000, 0x0?})
* 	GOROOT/src/crypto/tls/conn.go:1370 +0x156
* bufio.(*Reader).Read(0xc0139bdb00, {0xc0077c78c0, 0x9, 0x0?})
* 	GOROOT/src/bufio/bufio.go:241 +0x197
* io.ReadAtLeast({0x68de3a0, 0xc0139bdb00}, {0xc0077c78c0, 0x9, 0x9}, 0x9)
* 	GOROOT/src/io/io.go:335 +0x90
* io.ReadFull(...)
* 	GOROOT/src/io/io.go:354
* golang.org/x/net/http2.readFrameHeader({0xc0077c78c0, 0x9, 0xc00b79fd40?}, {0x68de3a0?, 0xc0139bdb00?})
* 	golang.org/x/net/http2/external/org_golang_x_net/http2/frame.go:237 +0x65
* golang.org/x/net/http2.(*Framer).ReadFrame(0xc0077c7880)
* 	golang.org/x/net/http2/external/org_golang_x_net/http2/frame.go:498 +0x85
* google.golang.org/grpc/internal/transport.(*http2Client).reader(0xc01399f8c8, 0xc0139bdb60)
* 	google.golang.org/grpc/internal/transport/external/org_golang_google_grpc/internal/transport/http2_client.go:1597 +0x22d
* created by google.golang.org/grpc/internal/transport.newHTTP2Client in goroutine 14903
* 	google.golang.org/grpc/internal/transport/external/org_golang_google_grpc/internal/transport/http2_client.go:394 +0x1c53
* 
* goroutine 16148 [select, 2 minutes]:
* google.golang.org/grpc.(*ClientConn).WaitForStateChange(0xc0072cf888, {0x69223d8, 0xc0183fd380}, 0x2)
* 	google.golang.org/grpc/external/org_golang_google_grpc/clientconn.go:515 +0xb1
* github.com/cockroachdb/cockroach/pkg/rpc.launchConnStateWatcher.func1({0x69223d8, 0xc0183fd380})
* 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/peer.go:868 +0x58
* github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
* 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:485 +0x13a
* created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 16143
* 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:476 +0x3fe
* 
* goroutine 51401 [select]:
* github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeWriteBatch(0xc007d5d108, {0x69223a0, 0xc013e7f4f0}, 0xc00b7466c0, 0xc0128ae9c0)
* 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_write.go:219 +0x807
* github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeBatchWithConcurrencyRetries(0xc007d5d108, {0x69223a0, 0xc013e7f4f0}, 0xc00b7466c0, 0x57e1c40)
* 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_send.go:512 +0x3c3
* github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).SendWithWriteBytes(0xc007d5d108, {0x69223a0?, 0xc013e7f4f0?}, 0xc00b7466c0)
* 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_send.go:186 +0x6e5
* github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).Send(0xc00b7466c0?, {0x69223a0?, 0xc013e7f4f0?}, 0x2000107?)
* 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_send.go:121 +0x1d
* github.com/cockroachdb/cockroach/pkg/kv/kvserver.sendProbe({0x69223a0, 0xc013e7f4f0}, {0x695f480, 0xc007d5d108})
* 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_circuit_breaker.go:228 +0x152
* github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaCircuitBreaker).asyncProbe.func1({0x69223d8?, 0xc012dea3f0?})
* 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_circuit_breaker.go:201 +0xcf
* github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
* 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:485 +0x13a
* created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 51400
* 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:476 +0x3fe
* 
*
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=23</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #123537 cli/democluster: TestTransientClusterSimulateLatencies failed [C-test-failure O-robot P-2 T-sql-foundations branch-master]
</p>
</details>
/cc @cockroachdb/sql-foundations @cockroachdb/server @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTransientClusterSimulateLatencies.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-38759

--- a/pkg/cli/democluster/demo_cluster_test.go
+++ b/pkg/cli/democluster/demo_cluster_test.go
@@ -173,7 +173,8 @@ func TestTransientClusterSimulateLatencies(t *testing.T) {
 		}
 	}()
 
-	ctx := context.Background()
+	ctx, cancelCtx := context.WithCancel(context.Background())
+	defer cancelCtx()
 
 	// Setup the transient cluster.
 	c := transientCluster{
@@ -185,16 +186,16 @@ func TestTransientClusterSimulateLatencies(t *testing.T) {
 		warnLog:           log.Warningf,
 		shoutLog:          log.Ops.Shoutf,
 	}
+
+	// Ensure the context gets canceled when the stopper terminates.
+	ctx, _ = c.stopper.WithCancelOnQuiesce(ctx)
+
+	require.NoError(t, c.Start(ctx))
+
 	// Stop the cluster when the test exits, including when it fails.
 	// This also calls the Stop() method on the stopper, and thus
 	// cancels everything controlled by the stopper.
 	defer c.Close(ctx)
-
-	// Also ensure the context gets canceled when the stopper
-	// terminates above.
-	ctx, _ = c.stopper.WithCancelOnQuiesce(ctx)
-
-	require.NoError(t, c.Start(ctx))
 
 	c.SetSimulatedLatency(true)
 

